File: docs/examples/sdks/python/adapters.py
from hushspec import HushGuard, HushSpecDenied, EvaluationAction
from hushspec.adapters import hush_tool, secure_tool
import sys

guard = HushGuard.from_file(sys.argv[1] if len(sys.argv) > 1 else "policy.yaml")
dispatches = []


@hush_tool(guard, tool_name="deploy")
def deploy():
    dispatches.append("deploy")


@secure_tool(guard, tool_name="write_file")
def write_file(content: str):
    dispatches.append(content)


@hush_tool(guard, action_type="file_read",
           action_mapper=lambda args, kwargs: EvaluationAction(type="file_read", target=args[0]))
def read_file(path: str):
    dispatches.append(path)


for operation in (deploy, lambda: write_file("synthetic"), lambda: read_file("/workspace/.env")):
    try:
        operation()
    except HushSpecDenied:
        pass
    else:
        raise AssertionError("Denied tool reached its body")
assert dispatches == []
try:
    invalid = hush_tool(guard, action_type="file_read")(lambda path: dispatches.append(path))
    invalid("/workspace/file.txt")
except ValueError as error:
    assert "action_mapper" in str(error)
else:
    raise AssertionError("Missing action mapper was accepted")
assert dispatches == []
guard.close()
print("PASS: Python LangChain/CrewAI decorators block before dispatch and require explicit effect mapping")

File: docs/examples/sdks/typescript/adapters.mjs
import assert from 'node:assert/strict';
import { HushGuard, mapOpenAIToolCall, mapClaudeToolToAction, wrapLangChainTool } from '@hushspec/core';

const guard=HushGuard.fromFile(process.argv[2] ?? 'policy.yaml');
let dispatches=0;
function ownedHandler(action) {
  guard.enforce(action);
  dispatches += 1;
}
assert.throws(()=>ownedHandler(mapOpenAIToolCall('deploy','{}')));
assert.throws(()=>ownedHandler(mapClaudeToolToAction('text_editor_20250429',{command:'view',path:'/workspace/.env'})));
const tool={name:'deploy',invoke:async()=>{dispatches+=1;}};
await assert.rejects(()=>wrapLangChainTool(tool,guard).invoke());
assert.equal(dispatches,0);
ownedHandler(mapOpenAIToolCall('search','{"query":"synthetic"}'));
assert.equal(dispatches,1);
console.log('PASS: structural OpenAI, Anthropic and LangChain dispatch boundaries');
